pandas/core/internals/managers.py

```diff
@@ -817,6 +818,15 @@ def get_slice(self, slobj: slice, axis: int = 0) -> BlockManager:
 
         return type(self)._simple_new(tuple(new_blocks), new_axes)
 
+    def get_slice_index(self, slobj: slice) -> BlockManager:
```

Comment:
can you add a get_slice_columns and change get_slice to just dispatch? (to share code)

Reply:
the only code-sharing i see available would be to change L819-L819 to `return self.get_slice_index(slobj)`which only saves 1 line

Flip side: one of the things im looking at is making another attempt at getting `axes` out of BlockManager (and ArrayManager), which would marginally widen the scope for sharing

Reply:
What does this specialized version avoid? Isn't it only the `if axis == 0:`/ `elif axis == 1` checks? 
(which I assume are quite fast checks, although if done repeatedly might indeed still worth to avoid)

The other difference is that `get_slice` does

```
            slicer = (slice(None), slobj)
            new_blocks = [blk.getitem_block(slicer) for blk in self.blocks]
```

while this new function does:

```
new_blocks = [blk.getitem_block_index(slobj) for blk in self.blocks]
```

but `get_slice` could do that as well?

Reply:
Can you answer this?

Reply:
>  Isn't it only the if axis == 0:/ elif axis == 1 checks?  (which I assume are quite fast checks, although if done repeatedly might indeed still worth to avoid)

Yes.  The code I'm profiling to optimize is (https://github.com/pandas-dev/pandas/pull/40171#issuecomment-790219422)

```
import numpy as np
import pandas as pd

N = 10 ** 4
labels = np.random.randint(0, 2000, size=N)
labels2 = np.random.randint(0, 3, size=N)
df = pd.DataFrame(
    {
        "key": labels,
        "key2": labels2,
        "value1": np.random.randn(N),
        "value2": ["foo", "bar", "baz", "qux"] * (N // 4),
    }
)

%prun -s cumtime for i in range(100): df.groupby("key").apply(lambda x: 1)  # on master as of this writing

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    4.170    4.170 {built-in method builtins.exec}
        1    0.001    0.001    4.170    4.170 <string>:1(<module>)
      100    0.001    0.000    4.156    0.042 groupby.py:911(apply)
      100    0.004    0.000    4.150    0.042 groupby.py:960(_python_apply_general)
      100    0.497    0.005    4.084    0.041 ops.py:263(apply)
   199000    0.176    0.000    3.223    0.000 ops.py:1005(__iter__)
   198900    0.248    0.000    2.986    0.000 ops.py:1046(_chop)
   198900    0.339    0.000    2.522    0.000 managers.py:796(get_slice)
   198900    0.188    0.000    1.631    0.000 managers.py:803(<listcomp>)
   596700    0.823    0.000    1.443    0.000 blocks.py:324(getitem_block)
   198900    0.135    0.000    0.409    0.000 base.py:4638(_getitem_slice)
```

we're still spending 8% of the runtime in get_slice itself, which is why i think avoiding re-doing these checks is worthwhile.

> but get_slice could do that as well?

sure, could update get_slice to also use `getitem_block_index` (i think the first draft of this PR started before `get_slice` reliably actually got a slice object)

Reply:
> we're still spending 8% of the runtime in get_slice itself, which is why i think avoiding re-doing these checks is worthwhile.

And did that number reduce by using get_slice_index instead?

> sure, could update get_slice to also use getitem_block_index

Yes, let's at least make the code consistent there

Reply:
Using this branch, timing `mgr.get_slice_index(slobj)` vs `mgr.get_slice(slobj, axis=1)` (but making them both use `getitem_block_index`, and removing the assert) , I only get a tiny difference (less than 5% difference, and that's then on something that only takes 8% of the total runtime)

Reply:
im seeing 17-22% gains with just getitem_block_index and not the extra BM/AM methods, so will revert those for this PR